flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java

```diff
@@ -144,6 +147,146 @@ public static Operation convertChangeColumn(
+
+        // generate new schema
+        return new AlterTableSchemaOperation(
+                tableIdentifier,
+                CatalogTable.of(
+                        builder.build(),
+                        catalogTable.getComment(),
+                        newPartitionKeys,
+                        catalogTable.getOptions()));
+    }
+
+    private static void validateColumnName(
+            String originColumnName,
+            String newColumnName,
+            List<String> columnNames,
+            ResolvedSchema originResolvedSchema) {
+        int originColumnIndex = columnNames.indexOf(originColumnName);
+        if (originColumnIndex < 0) {
+            throw new ValidationException(
+                    String.format("Old column %s not found for RENAME COLUMN ", originColumnName));
+        }
+
+        int sameColumnNameIndex = columnNames.indexOf(newColumnName);
+        if (sameColumnNameIndex >= 0) {
+            throw new ValidationException(
+                    String.format("New column %s existed for RENAME COLUMN ", newColumnName));
+        }
+
+        // validate old column name is referenced by computed column case
+        originResolvedSchema.getColumns().stream()
+                .filter(column -> column instanceof Column.ComputedColumn)
+                .forEach(
+                        column -> {
+                            Column.ComputedColumn computedColumn = (Column.ComputedColumn) column;
+                            Set<String> referencedColumn =
+                                    ColumnReferenceFinder.findReferencedColumn(
+                                            computedColumn.getExpression(), columnNames);
+                            if (referencedColumn.contains(originColumnName)) {
+                                throw new ValidationException(
+                                        String.format(
+                                                "Old column %s is referenced by computed column %s, currently doesn't "
+                                                        + "allow to rename column which is referenced by computed column.",
+                                                originColumnName,
+                                                computedColumn.asSummaryString()));
+                            }
+                        });
+    }
+
+    private static void buildNewColumnFromOriginColumn(
+            Schema.Builder builder, Schema.UnresolvedColumn originColumn, String columnName) {
+        if (originColumn instanceof Schema.UnresolvedComputedColumn) {
+            builder.columnByExpression(
+                    columnName, ((Schema.UnresolvedComputedColumn) originColumn).getExpression());
+        } else if (originColumn instanceof Schema.UnresolvedPhysicalColumn) {
+            builder.column(
+                    columnName, ((Schema.UnresolvedPhysicalColumn) originColumn).getDataType());
+        } else {
```

Comment:
Nit: is it will be better do the `column instance of`   again in here intead of assuming all the other column is instance of `UnresolvedMetadataColumn` in case of new types of Column will be added in the future?

Reply:
good catch, you are right